### PR TITLE
Improvement/skyline

### PR DESCRIPTION
**Pull request by bneumann**
Here we are. The SkyBottomLineCalculator in all it's beauty. Please take your time and review as I refactored a whole bunch of stuff.
This PR will automatically merge itself in 2 weeks :)

### CODE REVIEW

**Comment by bneumann** [comment]
These failing tests worry me deeply :(

**Comment by sebastianhaas** [comment]
How can I take a look at this? How can I activate the skyline rendering? :)

**Comment by bneumann** [comment]
Checkout the feature branch and hit npm start. Then check the demo page. Should be self explanatory

**Comment by sebastianhaas** [comment]
Looks great :) 

I'm experiencing a few minor problems that go away with different zoom levels- but in fact I can't really add to the discussion here. Just for reference: 

grafik

I also found that the debug shapes are only rendered for the top voice? Is that intended?


The only thing that I would really ask for is whether we could squash merge this into one commit with commit message in the shape of

```
feat: added sky and bottom lines
```
or
```
feat: added debugging helpers
```
or
```
feat(rendering): added sky and bottom lines
```
or
```
feat(debugging): added sky and bottom lines
```
or whatever makes sense as component or area, so that we don't need to search through 25 commits for the change log :)

**Comment by bneumann** [comment]
Squash an merge is fine. That's why I put it in a PR. Comment looks fine. Pick one 😁

Shapes: yes there are some boxes missing. Still investigating why. Possible missing initialization or drawing error. 

Skyline: the low precision is known and currently not avoidable. We also need to have a look at the performance too and whilst doing so fix the former too.

**Comment by sebastianhaas** [comment]
> Shapes: yes there are some boxes missing. Still investigating why. Possible missing initialization or drawing error.

That's curious, as far as I can tell it's really just the first voice for me that is being rendered with debug symbols. Is this a show-stopper for completing this PR?

Performance is OK on my machine, although it takes up to 2 seconds to initially calculate them.

**Comment by bneumann** [comment]
Nah, the boxes are no showstopper. And with performance I meant the Skyline. The boxes are only for debugging so doesn't matter

**Comment by sebastianhaas** [comment]
I can make a release after this is merged if you want for #285 etc to be fixed effectively.

**Comment by bneumann** [comment]
Sounds good but I really want to let this ripe a bit in the develop branch to check for side effects before releasing

**Comment by bneumann** [comment]
Meh. Now it already became stall.

File: src/MusicalScore/Graphical/GraphicalMeasure.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ import {PointF2D} from "../../Common/DataObjects/PointF2D";
  */
 export abstract class GraphicalMeasure extends GraphicalObject {
     protected firstInstructionStaffEntry: GraphicalStaffEntry;
-    protected lastInstructionStaffEntry: GraphicalStaffEntry;
+    public lastInstructionStaffEntry: GraphicalStaffEntry;
```
Comment (bneumann) [inline src/MusicalScore/Graphical/GraphicalMeasure.ts]: Can be removed, there are Getter/Setter for this

File: src/MusicalScore/Graphical/GraphicalMusicSheet.ts [hunk 2]
```diff
@@ -55,6 +55,8 @@ export class GraphicalMusicSheet {
     //private systemImages: Dictionary<MusicSystem, SystemImageProperties> = new Dictionary<MusicSystem, SystemImageProperties>();
     private numberOfStaves: number;
     private leadSheet: boolean = false;
+    private skyLineVisible: boolean = false;
```
Comment (bneumann) [inline src/MusicalScore/Graphical/GraphicalMusicSheet.ts]: Out of sheet-> only in drawer

File: src/MusicalScore/Graphical/MusicSheetCalculator.ts [hunk 3]
```diff
@@ -223,6 +257,42 @@ export abstract class MusicSheetCalculator {
...
+     */
+    protected optimizeDistanceBetweenStaffLines(musicSystem: MusicSystem): void {
+        musicSystem.PositionAndShape.calculateAbsolutePositionsRecursive(0, 0);
+
+        // don't perform any y-spacing in case of a StaffEntryLink (in both StaffLines)
+        if (!musicSystem.checkStaffEntriesForStaffEntryLink()) {
+            for (let i: number = 0; i < musicSystem.StaffLines.length - 1; i++) {
+                // const upperBottomLine: number = Math.max(...musicSystem.StaffLines[i].BottomLine);
+                const upperBottomLine: number = musicSystem.StaffLines[i].SkyBottomLineCalculator.getBottomLineMax();
+                // const lowerSkyLine: number = Math.min(...musicSystem.StaffLines[i + 1].SkyLine);
```
Comment (bneumann) [inline src/MusicalScore/Graphical/MusicSheetCalculator.ts]: document why lowerSkyLine is not used

File: src/MusicalScore/Graphical/MusicSheetCalculator.ts [hunk 4]
```diff
@@ -1463,18 +1489,31 @@ export abstract class MusicSheetCalculator {
...
+    //     /* tslint:enable:no-unused-variable */
+    //     let graphicalStaffEntry: GraphicalStaffEntry;
+    //     graphicalStaffEntry = MusicSheetCalculator.symbolFactory.createStaffEntry(openTie.Start.ParentStaffEntry, measure);
+    //     graphicalStaffEntry.relInMeasureTimestamp = Fraction.minus(absoluteTimestamp, measure.parentSourceMeasure.AbsoluteTimestamp);
+    //     this.resetYPositionForLeadSheet(graphicalStaffEntry.PositionAndShape);
+    //     measure.addGraphicalStaffEntryAtTimestamp(graphicalStaffEntry);
+    //     return graphicalStaffEntry;
+    // }
+
+    // private updateSkyBottomLines(): void {
```
Comment (bneumann) [inline src/MusicalScore/Graphical/MusicSheetCalculator.ts]: Remove comments

File: src/MusicalScore/Graphical/MusicSheetDrawer.ts [hunk 5]
```diff
@@ -454,8 +463,10 @@ export abstract class MusicSheetDrawer {
                                                          (relBoundingRect.width + 0), (relBoundingRect.height + 0));
             tmpRect = this.applyScreenTransformationForRect(tmpRect);
             this.renderRectangle(tmpRect, <number>GraphicalLayers.Background, layer, 0.5);
-            this.renderLabel(new GraphicalLabel(new Label(dataObjectString), 1.2, TextAlignment.CenterCenter),
-                             layer, tmpRect.width, tmpRect.height, tmpRect.height, new PointF2D(tmpRect.x, tmpRect.y + 12));
+            if (type === "all") {
```
Comment (bneumann) [inline src/MusicalScore/Graphical/MusicSheetDrawer.ts]: Later over configuration

File: src/MusicalScore/Graphical/MusicSheetDrawer.ts [hunk 6]
```diff
@@ -454,8 +463,10 @@ export abstract class MusicSheetDrawer {
...
+                this.renderLabel(new GraphicalLabel(new Label(dataObjectString), 1.2, TextAlignment.CenterCenter),
```
Comment (bneumann) [inline src/MusicalScore/Graphical/MusicSheetDrawer.ts]: As small as possible, then this "filter" is not needed

File: src/MusicalScore/Graphical/SkyBottomLineCalculator.ts [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+        // Create a temporary canvas outside the DOM to draw the measure in.
+        const tmpCanvas: any = new CanvasVexFlowBackend();
+        // search through all Measures
+        for (const measure of this.StaffLineParent.Measures as VexFlowMeasure[]) {
+            // must calculate first AbsolutePositions
+            measure.PositionAndShape.calculateAbsolutePositionsRecursive(0, 0);
+
+            // Pre initialize and get stuff for more performance
+            const vsStaff: any = measure.getVFStave();
+            tmpCanvas.initializeHeadless(vsStaff.getWidth());
```
Comment (bneumann) [inline src/MusicalScore/Graphical/SkyBottomLineCalculator.ts]: Add comment, that we are not in DOM

File: src/MusicalScore/Graphical/SkyBottomLineCalculator.ts [hunk 8]
```diff
@@ -0,0 +1,498 @@
...
+                        tmpBottomLine[x] = y;
+                        break;
+                    }
+                }
+            }
+            this.mSkyLine.push(...tmpSkyLine);
+            this.mBottomLine.push(...tmpBottomLine);
+
+            const debugTmpCanvas: boolean = false;
+            if (debugTmpCanvas) {
```
Comment (bneumann) [inline src/MusicalScore/Graphical/SkyBottomLineCalculator.ts]: VIsualize skyline for debugging

File: src/MusicalScore/Graphical/SkyBottomLineCalculator.ts [hunk 9]
```diff
@@ -0,0 +1,498 @@
...
+    }
+
+
+    //#region Private methods
+
+    /**
+     * Update an array value inside a range
+     * @param array Array to fill in the new value
+     * @param startIndex start index to begin with (default: 0)
+     * @param endIndex end index of array (default: array length)
```
Comment (bneumann) [inline src/MusicalScore/Graphical/SkyBottomLineCalculator.ts]: comment: endIndex is NOT included

File: src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetDrawer.ts [hunk 10]
```diff
@@ -75,9 +77,11 @@ export class VexFlowMusicSheetDrawer extends MusicSheetDrawer {
         for (const voiceID in measure.vfVoices) {
             if (measure.vfVoices.hasOwnProperty(voiceID)) {
                 const tickables: Vex.Flow.Tickable[] = measure.vfVoices[voiceID].tickables;
-                for (const tick of tickables) {
-                    if ((<any>tick).getAttribute("type") === "StaveNote" && process.env.DEBUG) {
-                        tick.getBoundingBox().draw(this.backend.getContext());
+                if (process.env.DEBUG) {
```
Comment (bneumann) [inline src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetDrawer.ts]: This can be removed

**Pull request merged**